src/hal/components/limit3.comp

```diff
@@ -2,94 +2,118 @@ component limit3 "Limit the output signal to fall between min and max, limit its
 pin in float in;
 pin out float out;
 pin in bit load "When TRUE, immediately set \\fBout\\fB to \\fBin\\fR, ignoring maxv and maxa";
-pin in float min_=-1e20;
-pin in float max_=1e20;
+pin in float min=-1e20;
+pin in float max=1e20;
 pin in float maxv=1e20;
 pin in float maxa=1e20;
-option data limit3_data;
+variable double in_pos_old;
+variable double out_vel_old;
 function _;
 license "GPL";
 ;;
 
 #include "rtapi_math.h"
 
-typedef struct {
-    double old_in;	/* previous input */
-    double old_out;	/* previous output */
-    double old_v;	/* previous 1st derivative */
-} limit3_data;
+#define SET_NEXT_STATE(out_pos, out_vel, in_pos)	\
+    do {						\
+	out = out_pos;					\
+	out_vel_old = out_vel;				\
+	in_pos_old = in_pos;				\
+	return;						\
+    } while (0)
 
-FUNCTION(_) {
-    double lin, lout, dt, in_v, min_v, max_v, ramp_a, avg_v, err, dv, dp;
-    double min_out, max_out, match_time, est_in, est_out;
+#define VALID_NEXT(pos, vel) ((pos) <= max_pos && (pos) >= min_pos	\
+			      && (vel) <= max_vel && (vel) >= min_vel)
 
-    /* apply first order limit */
-    lin = in;
-    if ( lin < min_ ) {
-	lin = min_;
-    }
-    if ( lin > max_ ) {
-	lin = max_;
-    }
+FUNCTION(_) {
+    double in_pos_lim, dt, in_vel, min_vel, max_vel, min_pos, max_pos;
+    double vel_0_time, vel_0_pos;
+    double vel_match_time, vel_match_in_pos, vel_match_out_pos;
+    int out_dir, out_dir_rel;
 
-    if(load) {
-	data.old_in = data.old_out = out = lin;
-	data.old_v = 0;
+    if (load) {
+	// Apply first order limit
+	in_pos_lim = fmin(max, fmax(min, in));
+	SET_NEXT_STATE(in_pos_lim, 0, in_pos_lim);
 	return;
     }
 
-    /* calculate input derivative */
-    dt = period * 0.000000001;
-    in_v = (lin - data.old_in) / dt;
-    /* determine v and out that can be reached in one period */
-    min_v = data.old_v - maxa * dt;
-    if ( min_v < -maxv ) {
-	min_v = -maxv;
+    // Time increment per update
+    dt = period * 1e-9;
```

Comment:
I think you can refer to 'fperiod' to get this computed value, but I'd have to double check comp docs.

Reply:
Cool, fixed.